CMaxTable.lua

```diff
@@ -4,25 +4,26 @@ function CMaxTable:__init()
    parent.__init(self)
    self.gradInput = {}
    self.maxIdx = torch.Tensor()
+   self.mask = torch.Tensor() -- reused for memory allocation efficiency
 end
 
 function CMaxTable:updateOutput(input)
    self.output:resizeAs(input[1]):copy(input[1])
    self.maxIdx:resizeAs(input[1]):fill(1)
    for i=2,#input do
-      local mask = torch.gt(input[i], self.output)
-      self.maxIdx:maskedFill(mask, i)
-      self.output:maskedCopy(mask, input[i][mask])
+      self.mask = torch.gt(input[i], self.output)
```

Comment:
This still allocates memory at every `forward` run, enforcing synchronization points from time to time when the memory is released.
If you could do instead something like
```lua
input[i].gt(self.mask, input[i], self.output)
```
you will avoid memory allocations and it should be faster.